### PR TITLE
policymap, fragmap: clean up doc, map size configuration

### PR DESCRIPTION
**Pull request by qmonnet**
More clean-up following the introduction of IPv4 fragment tracking.

Follow-up on #10927.

First commit updates the doc for kubeproxy-free (removes note about missing fragmentation support).

Second commit cleans up the way we configure map size for policymap and fragmap, and falls back to limit values if the user passes unreasonable values. No lower limit for fragmap (other than 0), maybe we could remove it for policymap as well but I'm unsure if the lower limit is used elsewhere as an assumption. Anyway it's already possible to ask for as few as 256 entries in policy map, not sure it's worth trying to allow fewer than that.

### CODE REVIEW

**Comment by qmonnet** [comment]
test-me-please

File: pkg/defaults/defaults.go [hunk 1]
```diff
@@ -246,6 +246,14 @@ const (
 	// the endpoint could send/receive traffic to/from.
 	PolicyMapEntries = 16384 // Cilium 1.5 and earlier value
 
+	// PolicyMapMinSize is the minimum value to which the number of entries
```
Comment (tklauser) [inline pkg/defaults/defaults.go]: This will conflict with the refactoring in #10957 (for #10780) where these consts are moved to the top of `pkg/option/config` (like the other map min/max constants). Same for `FragmentMapMaxSize`. I think we can keep them in `pkg/option` as they are not used outside of the package.

I think it makes sense to keep the default map sizes in `pkg/defaults` though.

**Review by pchaigno — APPROVED** [review]
LGTM!

File: pkg/defaults/defaults.go (hunk 1 above)
Comment (qmonnet) [reply pkg/defaults/defaults.go]: With your change they would be in a list of similar limits for the other maps, which looks better than leaving them in the middle of the `config` package as it is now, so OK.

I'll probably hold until your PR is merged in that case, and rebase afterwards.

Thanks!

**Comment by qmonnet** [comment]
No more functional change in second commit, I'm good with the clean-up from Tobias on map size limits. Which makes this set smaller, and pure clean-up.

Note the change on `policymap.MaxEntries` initialisation.

test-me-please

**Comment by tklauser** [comment]
test-me-please

**Review by tklauser — APPROVED** [review]
Thanks, LGTM!

**Review by pchaigno — APPROVED** [review]

**Comment by qmonnet** [comment]
restart-ginkgo

**Comment by qmonnet** [comment]
Ok, so removing initialisation of `policymap.MaxEntries` seemed like a good idea, but `eppolicymap.CreateEPPolicyMap()` uses the initial value for its unit tests. I set it back to its initial value, but by calling the const in `defaults`.

It turns out we got some nasty nil pointer dereference in the CI, because `eppolicymap.CreateEPPolicyMap()` doesn't care if BPF map creation fails and returns a nil pointer, and keeps going with `OpenOrCreate()` (where we try to dereference the map). So I added a third commit to fix it.

**Comment by qmonnet** [comment]
test-me-please

**Comment by qmonnet** [comment]
Ginkgo / Unit tests for eppolicymap fail because of my change:
```
[2020-04-15T15:55:55.170Z]   	 START: eppolicymap_test.go:54: EPPolicyMapTestSuite.TestCreateEPPolicy
[2020-04-15T15:55:55.170Z]   	 START: eppolicymap_test.go:42: EPPolicyMapTestSuite.SetUpTest
[2020-04-15T15:55:55.170Z]   	 PASS: eppolicymap_test.go:42: EPPolicyMapTestSuite.SetUpTest	0.000s
[2020-04-15T15:55:55.170Z]   	 
[2020-04-15T15:55:55.170Z]   	 level=info msg="Detected mounted BPF filesystem at /sys/fs/bpf" subsys=bpf
[2020-04-15T15:55:55.170Z]   	 level=warning msg="Unable to open or create endpoint policy map" error="Unable to create map /sys/fs/bpf/tc/globals/unit_test_ep_to_policy: invalid argument" subsys=map-ep-policy
[2020-04-15T15:55:55.170Z]   	 eppolicymap_test.go:57:
[2020-04-15T15:55:55.170Z]   	     c.Assert(err, IsNil)
[2020-04-15T15:55:55.170Z]   	 ... value *os.PathError = &os.PathError{Op:"Unable to create map", Path:"/sys/fs/bpf/tc/globals/unit_test_ep_to_policy", Err:0x16} ("Unable to create map /sys/fs/bpf/tc/globals/unit_test_ep_to_policy: invalid argument")
[...]
```
The asserts I added do not pass. Not sure what happens exactly, the map creation succeeds on my setup and I can't reproduce the error locally. Will need more investigation.

(Kernel tests: Istio tests `could not find pods in namespace 'kube-system' with label 'k8s-app=cilium': signal: killed`, flake?)

**Review by pchaigno — CHANGES_REQUESTED** [review]
Actually, could you also fix up [the sentences in `config.go`](https://github.com/cilium/cilium/blob/1000f7023a46f2ff0c29b3b155c294942b5c4fd7/pkg/option/config.go#L2752)? That succession of compound adjectives without any punctuation is kinda hard to parse.

**Comment by qmonnet** [comment]
My fix for the `eppolicymap` unit tests was incorrect. I updated the commit (and also unset the value from `policy.MaxEntries`, which no longer needs be set even for `eppolicymap` unit tests).

**Comment by qmonnet** [comment]
test-me-please

**Comment by qmonnet** [comment]
`eppolicymap` unit tests do not need an initial value for `policymap.MaxEntries`, but `policymap` unit tests do -_-. I'm initialising again at `defaults.PolicyMapEntries`. This time we should be good :crossed_fingers:.

**Comment by qmonnet** [comment]
test-me-please

**Comment by qmonnet** [comment]
test-me-please

**Review by pchaigno — APPROVED** [review]
Nice catch on the null dereference for eppolicymap!
That will require a backport to 1.7, no?

**Comment by qmonnet** [comment]
ah, right, thanks!

File: pkg/maps/eppolicymap/eppolicymap.go [hunk 2]
```diff
@@ -60,6 +60,18 @@ var (
 // for testing purposes.
 func CreateWithName(mapName string) error {
 	buildMap.Do(func() {
+		EpPolicyMap = bpf.NewMap(mapName,
```
Comment (tgraf) [inline pkg/maps/eppolicymap/eppolicymap.go]: Maybe instead of `Warning()` it should be `Fatal()`. What's the point in continuing if `fd` is 0?

**Review by tgraf — DISMISSED** [review]

File: pkg/maps/eppolicymap/eppolicymap.go (hunk 2 above)
Comment (qmonnet) [reply pkg/maps/eppolicymap/eppolicymap.go]: Right, it's probably useless to carry on indeed. I just changed to `Fatal()` and respined. Thanks!

**Comment by qmonnet** [comment]
test-me-please

**Comment by aanm** [comment]
@qmonnet please rebase your PR since timeouts are fixed in master.

**Comment by qmonnet** [comment]
test-me-please

**Comment by qmonnet** [comment]
test-gke

**Comment by mrostecki** [comment]
To backport this pull request to 1.7, we need also backport #10626, am I right? @qmonnet @pchaigno

**Comment by pchaigno** [comment]
@mrostecki I think only commit `eppolicymap: exit if inner map creation fails, to avoid nil dereference` requires a backport here and it should be independent from my previous PR :-)

**Comment by mrostecki** [comment]
@pchaigno OK, backporting that one commit has no conflicts. Thanks a lot!
